Lazy Imports not getting resolved correctly in bundles

**Steps to Reproduce:**

1. Clone https://github.com/jay8t6/lazy-import-bundle
2. Run `npm i;bower i`
3. Run `npm run build`
4. Run `polymer serve build/ -o`
5. Open console

**Expected Result:** `paper-button` will lazily import into the side bar

**Actual Result:** `paper-button` doesn't import. You can see in the console that the import failed and resolved to `bower_components/bower_components/paper-button/paper-button.html`

@usergenic

Added a failing test to prove the behavior in https://github.com/Polymer/polymer-bundler/tree/lazy-import-href-resolution-investigation
```
  1) Bundler lazy imports are treated like entrypoints:
     AssertionError: The href of the lazy import link in the inlined dom-module should not be modified: expected '<div hidden="" by-polymer-bundler=""><link rel="import" href="../shared_bundle_1.html"><l
ink rel="import" href="shared-eager-and-lazy-import-1.html"><dom-module id="eager-import-1" assetpath="subfolder/">\n  <link type="lazy-import" href="subfolder/lazy-import-3.html">\n  <script>\n    conso
le.log(\'eager-import-1\');\n  </script>\n</dom-module>\n</div><div id="lazy-import-2">\n  I am lazy import 2\n</div>\n' to include '<link rel="lazy-import" href="lazy-import-3.html">'
```

Updating bundler code to stop monkeying with hrefs inside dom-modules that have been inlined.

A little concerned this issue might be related to more than just lazy imports.  I don't know if assetpath is honored for any urls other than lazy ones, but keeping an eye out for them.

Anyways, hopefully will have a reasonably straightforward fix for this in a moment.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,9 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/)
 and this project adheres to [Semantic Versioning](http://semver.org/).
 
-<!-- ## Unreleased -->
-
+## Unreleased
+- Fixed issue where an inlined import's `<link rel="lazy-import">` href
+  was being adjusted, but should not be adjusted when governed by the
+  containing dom-module's assetpath.
 <!-- Add new, unreleased changes here. -->
+
 
 ## 2.0.0-pre.15 - 2017-05-08
 - Fixed case where an inlined import's `<link rel="lazy-import">` tags

--- a/src/import-utils.ts
+++ b/src/import-utils.ts
@@ -363,7 +363,7 @@ function rewriteElementAttrsBaseUrl(
     rewriteUrlsInTemplates?: boolean) {
   const nodes = dom5.queryAll(
       ast,
-      matchers.urlAttrs,
+      matchers.elementsWithUrlAttrsToRewrite,
       undefined,
       rewriteUrlsInTemplates ? dom5.childNodesIncludeTemplate :
                                dom5.defaultChildNodes);

--- a/src/matchers.ts
+++ b/src/matchers.ts
@@ -90,8 +90,8 @@ export const hiddenDiv: Matcher = predicates.AND(
 export const inHiddenDiv: Matcher = predicates.parentMatches(hiddenDiv);
 
 export const elementsWithUrlAttrsToRewrite: Matcher = predicates.AND(
-    predicates.OR.apply(
-        null, constants.URL_ATTR.map((attr) => predicates.hasAttr(attr))),
+    predicates.OR(
+        ...constants.URL_ATTR.map((attr) => predicates.hasAttr(attr))),
     predicates.NOT(predicates.AND(
         predicates.parentMatches(predicates.hasTagName('dom-module')),
         lazyHtmlImport)));

--- a/src/matchers.ts
+++ b/src/matchers.ts
@@ -20,12 +20,6 @@ import {predicates} from 'dom5';
 import * as parse5 from 'parse5';
 
 export interface Matcher { (node: parse5.ASTNode): boolean; }
-;
-
-export const urlAttrMatchers: (Matcher)[] =
-    constants.URL_ATTR.map((attr) => predicates.hasAttr(attr));
-
-export const urlAttrs: Matcher = predicates.OR.apply(null, urlAttrMatchers);
 
 export const jsMatcher: Matcher = predicates.AND(
     predicates.hasTagName('script'),
@@ -66,16 +60,6 @@ export const externalJavascript: Matcher =
     predicates.AND(predicates.hasAttr('src'), jsMatcher);
 export const inlineJavascript: Matcher =
     predicates.AND(predicates.NOT(predicates.hasAttr('src')), jsMatcher);
-export const htmlImport: Matcher = predicates.AND(
-    predicates.hasTagName('link'),
-    predicates.OR(
-        predicates.hasAttrValue('rel', 'import'),
-        predicates.hasAttrValue('rel', 'lazy-import')),
-    predicates.hasAttr('href'),
-    predicates.OR(
-        predicates.hasAttrValue('type', 'text/html'),
-        predicates.hasAttrValue('type', 'html'),
-        predicates.NOT(predicates.hasAttr('type'))));
 export const eagerHtmlImport: Matcher = predicates.AND(
     predicates.hasTagName('link'),
     predicates.hasAttrValue('rel', 'import'),
@@ -84,6 +68,16 @@ export const eagerHtmlImport: Matcher = predicates.AND(
         predicates.hasAttrValue('type', 'text/html'),
         predicates.hasAttrValue('type', 'html'),
         predicates.NOT(predicates.hasAttr('type'))));
+export const lazyHtmlImport: Matcher = predicates.AND(
+    predicates.hasTagName('link'),
+    predicates.hasAttrValue('rel', 'lazy-import'),
+    predicates.hasAttr('href'),
+    predicates.OR(
+        predicates.hasAttrValue('type', 'text/html'),
+        predicates.hasAttrValue('type', 'html'),
+        predicates.NOT(predicates.hasAttr('type'))));
+export const htmlImport: Matcher =
+    predicates.OR(eagerHtmlImport, lazyHtmlImport);
 export const stylesheetImport: Matcher = predicates.AND(
     predicates.hasTagName('link'),
     predicates.hasAttrValue('rel', 'import'),
@@ -94,6 +88,13 @@ export const hiddenDiv: Matcher = predicates.AND(
     predicates.hasAttr('hidden'),
     predicates.hasAttr('by-polymer-bundler'));
 export const inHiddenDiv: Matcher = predicates.parentMatches(hiddenDiv);
+
+export const elementsWithUrlAttrsToRewrite: Matcher = predicates.AND(
+    predicates.OR.apply(
+        null, constants.URL_ATTR.map((attr) => predicates.hasAttr(attr))),
+    predicates.NOT(predicates.AND(
+        predicates.parentMatches(predicates.hasTagName('dom-module')),
+        lazyHtmlImport)));
 
 /**
  * TODO(usergenic): From garlicnation's PR comment - "This matcher needs to deal

--- a/src/test/bundler_test.ts
+++ b/src/test/bundler_test.ts
@@ -184,19 +184,18 @@ suite('Bundler', () => {
     const manifest = await bundler.generateManifest(['lazy-imports.html']);
     const documents = await bundler.bundle(manifest);
 
-    // The `lazy-imports.html` file has 3 imports in the head of the
+    // The `lazy-imports.html` file has 2 imports in the head of the
     // document.  The first is eager and should be moved.  The remaining
-    // two are lazy imports and should not be moved.
+    // one is lazy and should not be moved.
     const entrypointBundle = documents.get('lazy-imports.html')!.ast;
-    console.log(parse5.serialize(entrypointBundle));
     const entrypointLazyImports = dom5.queryAll(
         entrypointBundle,
         // Query for all HTML imports so we can prove only the 2 lazy ones
         // remain.
         preds.AND(preds.parentMatches(matchers.head), matchers.htmlImport));
-    assert.equal(entrypointLazyImports.length, 2);
+    assert.equal(entrypointLazyImports.length, 1);
     assert.equal(dom5.getAttribute(entrypointLazyImports[0]!, 'group'), 'one');
-    assert.equal(dom5.getAttribute(entrypointLazyImports[1]!, 'group'), 'two');
+
 
     // The shared bundle has an inlined dom-module with an embedded
     // lazy-import via `shared-eager-import-2.html` that we are verifying
@@ -269,6 +268,32 @@ suite('Bundler', () => {
     const sharedEagerBundle =
         parse5.serialize(documents.get('shared_bundle_1.html')!.ast);
     assert.include(sharedEagerBundle, '<div id="shared-eager-import-2">');
+  });
+
+  test('lazy imports governed by dom-module assetpath', async () => {
+    const bundler = new Bundler({
+      analyzer: new Analyzer({urlLoader: new FSUrlLoader('test/html/imports')}),
+      // We exclude this import so we can examine how the eager import URL
+      // is modified in the bundled lazy-import-2.html document.
+      excludes: ['lazy-imports/subfolder/eager-import-2.html'],
+    });
+    const manifest = await bundler.generateManifest(['lazy-imports.html']);
+    const documents = await bundler.bundle(manifest);
+    const lazyImport2 =
+        parse5.serialize(documents.get('lazy-imports/lazy-import-2.html')!.ast);
+    assert.include(
+        lazyImport2,
+        '<dom-module id="eager-import-1" assetpath="subfolder/">',
+        'lazy-import-2.html should have inlined subfolder/eager-import-1.html');
+    assert.include(
+        lazyImport2,
+        '<link rel="lazy-import" href="lazy-import-3.html">',
+        'The href of the lazy import link in the inlined dom-module should not be modified');
+    assert.include(
+        lazyImport2,
+        '<link rel="import" href="subfolder/eager-import-2.html">',
+        'The href of an eager import link inside a dom-module should still be modified');
+
   });
 
   test('Handle <base> tag', async () => {

--- a/src/test/deps-index_test.ts
+++ b/src/test/deps-index_test.ts
@@ -63,9 +63,12 @@ suite('Bundler', () => {
       const entrypoint = 'lazy-imports.html';
       const lazyImport1 = 'lazy-imports/lazy-import-1.html';
       const lazyImport2 = 'lazy-imports/lazy-import-2.html';
+      const lazyImport3 = 'lazy-imports/subfolder/lazy-import-3.html';
       const shared1 = 'lazy-imports/shared-eager-import-1.html';
       const shared2 = 'lazy-imports/shared-eager-import-2.html';
       const shared3 = 'lazy-imports/shared-eager-and-lazy-import-1.html';
+      const eager1 = 'lazy-imports/subfolder/eager-import-1.html';
+      const eager2 = 'lazy-imports/subfolder/eager-import-2.html';
       const deeply1 = 'lazy-imports/deeply-lazy-import-1.html';
       const deeply2 = 'lazy-imports/deeply-lazy-imports-eager-import-1.html';
       const analyzer =
@@ -73,7 +76,11 @@ suite('Bundler', () => {
       const expectedEntrypointsToDeps = new Map([
         [entrypoint, new Set([entrypoint, shared3])],
         [lazyImport1, new Set([lazyImport1, shared1, shared2])],
-        [lazyImport2, new Set([lazyImport2, shared1, shared2, shared3])],
+        [
+          lazyImport2,
+          new Set([lazyImport2, shared1, shared2, shared3, eager1, eager2])
+        ],
+        [lazyImport3, new Set([lazyImport3])],
         [shared3, new Set([shared3])],
         [deeply1, new Set([deeply1, deeply2])],
       ]);

--- a/test/html/imports/lazy-imports.html
+++ b/test/html/imports/lazy-imports.html
@@ -1,12 +1,17 @@
 <html>
 
 <head>
-    <link rel="import" href="lazy-imports/shared-eager-and-lazy-import-1.html">
-    <link rel="lazy-import" group="one" href="lazy-imports/lazy-import-1.html">
-    <link rel="lazy-import" group="two" href="lazy-imports/lazy-import-2.html">
+  <link rel="import" href="lazy-imports/shared-eager-and-lazy-import-1.html">
+  <link rel="lazy-import" group="one" href="lazy-imports/lazy-import-1.html">
 </head>
 
 <body>
+  <dom-module id="element-1">
+    <link rel="lazy-import" group="two" href="lazy-imports/lazy-import-2.html">
+    <script>
+      console.log('element-1');
+    </script>
+  </dom-module>
 </body>
 
 </html>

--- a/test/html/imports/lazy-imports/lazy-import-1.html
+++ b/test/html/imports/lazy-imports/lazy-import-1.html
@@ -1,5 +1,8 @@
 <link rel="import" href="shared-eager-import-1.html">
-<link rel="lazy-import" href="shared-eager-and-lazy-import-1.html">
-<div id="lazy-import-1">
+
+<dom-module id="lazy-imported-element-1">
+  <link rel="lazy-import" href="shared-eager-and-lazy-import-1.html">
+  <template id="lazy-import-1">
     I am lazy import 1
-</div>
+  </template>
+</dom-module>

--- a/test/html/imports/lazy-imports/lazy-import-2.html
+++ b/test/html/imports/lazy-imports/lazy-import-2.html
@@ -1,5 +1,6 @@
 <link rel="import" href="shared-eager-import-1.html">
 <link rel="import" href="shared-eager-and-lazy-import-1.html">
+<link rel="import" href="subfolder/eager-import-1.html">
 <div id="lazy-import-2">
   I am lazy import 2
 </div>

--- a/test/html/imports/lazy-imports/subfolder/eager-import-1.html
+++ b/test/html/imports/lazy-imports/subfolder/eager-import-1.html
@@ -1,0 +1,7 @@
+<dom-module id="eager-import-1">
+  <link rel="import" href="eager-import-2.html">
+  <link rel="lazy-import" href="lazy-import-3.html">
+  <script>
+    console.log('eager-import-1');
+  </script>
+</dom-module>

--- a/test/html/imports/lazy-imports/subfolder/eager-import-2.html
+++ b/test/html/imports/lazy-imports/subfolder/eager-import-2.html
@@ -1,0 +1,3 @@
+<div id="eager-import-2">
+  Eager eager
+</div>

--- a/test/html/imports/lazy-imports/subfolder/lazy-import-3.html
+++ b/test/html/imports/lazy-imports/subfolder/lazy-import-3.html
@@ -1,0 +1,3 @@
+<div id="lazy-import-3">
+  Lazy lazy lazy
+</div>

